Packages/com.unity.inputsystem/InputSystem/Plugins/XR/XRTemplateBuilder.cs

```diff
@@ -0,0 +1,213 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Experimental.Input.LowLevel;
+using UnityEngine.Experimental.Input.Utilities;
+using UnityEngine.Experimental.Input.Plugins.XR.Haptics;
+using System.Text;
+
+namespace UnityEngine.Experimental.Input.Plugins.XR
+{
+    [Serializable]
+    class XRTemplateBuilder
+    {
+        static List<Func<XRDeviceDescriptor, string>> availableTemplates = new List<Func<XRDeviceDescriptor, string>>();
+
+        public XRDeviceDescriptor descriptor;
+
+        static uint GetOffsetForFeatureType(XRFeatureDescriptor featureDescriptor)
+        {
+            switch (featureDescriptor.featureType)
+            {
+                case EFeatureType.Binary:
+#if UNITY_ANDROID
+                        return 4;
+#else
+                    return 1;
+#endif
+                case EFeatureType.DiscreteStates:
+                    return sizeof(int);
+                case EFeatureType.Axis1D:
+                    return sizeof(float);
+                case EFeatureType.Axis2D:
+                    return sizeof(float) * 2;
+                case EFeatureType.Axis3D:
+                    return sizeof(float) * 3;
+                case EFeatureType.Rotation:
+                    return sizeof(float) * 4;
+                case EFeatureType.Custom:
+                    return featureDescriptor.customSize;
+            }
+            return 0;
+        }
+
+        public static void RegisterTemplateFilter(Func<XRDeviceDescriptor, string> templateChecker)
+        {
+            availableTemplates.Add(templateChecker);            
+        }
+
+        static string SanitizeTemplateName(string templateName)
+        {
+            int stringLength = templateName.Length;
+            StringBuilder sanitizedTemplateName = new StringBuilder(stringLength);
+            for (int i = 0; i < stringLength; i++)
+            {
+                char letter = templateName[i];
+                if(Char.IsUpper(letter) || Char.IsLower(letter) || Char.IsDigit(letter) || letter == ':')

```

Comment:
Reference primitive types with their primitive type name instead of their value struct name. I.e. `char` instead of `Char`.